src/ode/man/man3/uquit.3
.TH UQUIT 3 8/13/90
.CM 4
.SH NAME
uquit \- prints error message, program usage, then exits program
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" @OSF_FREE_COPYRIGHT@
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1992, 1991, 1990  
.\" Open Software Foundation, Inc. 
.\"  
.\" Permission is hereby granted to use, copy, modify and freely distribute 
.\" the software in this file and its documentation for any purpose without 
.\" fee, provided that the above copyright notice appears in all copies and 
.\" that both the copyright notice and this permission notice appear in 
.\" supporting documentation.  Further, provided that the name of Open 
.\" Software Foundation, Inc. ("OSF") not be used in advertising or 
.\" publicity pertaining to distribution of the software without prior 
.\" written permission from OSF.  OSF makes no representations about the 
.\" suitability of this software for any purpose.  It is provided "as is" 
.\" without express or implied warranty. 
.\"
.\" HISTORY
.\" $Log: uquit.3,v $
.\" Revision 1.4.2.2  1992/12/03  17:25:23  damon
.\" 	ODE 2.2 CR 183. Added CMU notice
.\" 	[1992/12/03  17:10:44  damon]
.\"
.\" Revision 1.4  1991/12/05  21:17:13  devrcs
.\" 	Added _FREE_ to copyright marker
.\" 	[91/08/01  08:17:42  mckeen]
.\" 
.\" Revision 1.3  90/10/07  21:59:11  devrcs
.\" 	Added EndLog Marker.
.\" 	[90/09/29  14:17:06  gm]
.\" 
.\" Revision 1.2  90/08/25  12:23:46  devrcs
.\" 	Initial revision
.\" 	[90/08/14  11:19:17  randyb]
.\" 
.\" $EndLog$
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH "SYNOPSIS"
.B "uquit ( status, usage, format [, arg ] ...)"
.br
.B "int status;"
.br
.B "int usage;"
.br
.B "char *format;"
.SH "DESCRIPTION"
.I
uquit
is a means of terminating a process with an error message
after having printed a usage message.
If the second argument, \fBusage\fR, is not 0,
\fBuquit\fR calls \fBprint_usage ()\fR.
.sp
A call to \fBuquit\fR should never return.
.sp
.I uquit
prints to standard error the message given in \fBformat\fR.
This is an argument list in the form specified by \fIprintf (3).\fR
\fBuquit\fR then exits with the value of \fBstatus\fR.
.sp
All of the process's files will be closed.
.SH "FILES"
.IR vfprintf (3), 
.IR exit (2), 
.IR wait (2)
.SH "EXIT VALUES"
The first argument,
.IR status ,
will be the argument passed to
.IR exit (2)
and this is the process's return code.


src/ode/lib/libode/uquit.c
/*
 * @OSF_FREE_COPYRIGHT@
 * COPYRIGHT NOTICE
 * Copyright (c) 1992, 1991, 1990  
 * Open Software Foundation, Inc. 
 *  
 * Permission is hereby granted to use, copy, modify and freely distribute 
 * the software in this file and its documentation for any purpose without 
 * fee, provided that the above copyright notice appears in all copies and 
 * that both the copyright notice and this permission notice appear in 
 * supporting documentation.  Further, provided that the name of Open 
 * Software Foundation, Inc. ("OSF") not be used in advertising or 
 * publicity pertaining to distribution of the software without prior 
 * written permission from OSF.  OSF makes no representations about the 
 * suitability of this software for any purpose.  It is provided "as is" 
 * without express or implied warranty. 
 */
/*
 * HISTORY
 * $Log: uquit.c,v $
 * Revision 1.6.5.5  1993/04/28  14:35:43  damon
 * 	CR 463. More pedantic changes
 * 	[1993/04/28  14:34:36  damon]
 *
 * Revision 1.6.5.4  1993/04/08  22:51:33  damon
 * 	CR 446. Clean up include files
 * 	[1993/04/08  22:51:16  damon]
 * 
 * Revision 1.6.5.3  1993/01/21  18:39:17  damon
 * 	CR 401. Added stdarg
 * 	[1993/01/21  18:34:12  damon]
 * 
 * Revision 1.6.2.2  1992/12/03  17:23:04  damon
 * 	ODE 2.2 CR 183. Added CMU notice
 * 	[1992/12/03  17:09:30  damon]
 * 
 * Revision 1.6  1991/12/05  21:13:28  devrcs
 * 	Changed sdm to ode; std_defs.h to  odedefs.h
 * 	[91/01/10  11:47:15  randyb]
 * 
 * 	Correct copyright; clean up lint input; added ui_print function.
 * 	[91/01/08  12:24:24  randyb]
 * 
 * 	Style update; saberc lint
 * 	[90/11/08  09:57:43  randyb]
 * 
 * Revision 1.4  90/10/07  20:05:00  devrcs
 * 	Pre-OSF/1 changes
 * 
 * $EndLog$
 */
/******************************************************************************
**                          Open Software Foundation                         **
**                              Cambridge, MA                                **
**                              Randy Barbano                                **
**                               April 1990                                  **
*******************************************************************************
**
**    Description:
**	This function is part of library libsb.a.  It purpose is to provide
**	a common method of exiting a program.  It provides a way of quitting
**	with or without calling print_usage.
**
**    lib functions and their usage:
**	1) uquit ( exit_value, usage, format, arg1, arg2... )
**	   args:
**	     int      exit_value;		-1, 0, 1, etc
**	     boolean  usage; 			true or false - 1 or 0
**	     char     format [],	 	"%s has value %d"
**	     	      args1 [], arg2 []...;	arguments to format
**
**	   returns:
**	     does not returns, calls exit with exit_value.
**
**	   usage:
**	     This procedure prints out an error message and exits the
**	     the program that called it.  It uses the form of ui_print to 
**	     output the message.  This program also depends on an external
**	     name, progname, to indicate which program was running when the
**	     failure occured.   It prints out the error message from format.
**	     Format is like a printf statement in that it can take a
**	     varying number of arguments.
**
**	     If usage is not set to 0, uquit will call the print_usage
**	     program before it exits with the value in exit_value.
**
**	     NOTE: print_usage is a routine called by uquit.  The
**	           user must create this procedure to use uquit.
*/

#include <stdio.h>
#include <ode/odedefs.h>
#include <ode/util.h>
#if __STDC__
#  include <stdarg.h>
#else
#  include <varargs.h>
#endif
    extern char   * progname;                             /* program running */

void
#if __STDC__
uquit ( int status, ... )
#else
uquit ( va_alist )
va_dcl
#endif
	/* This procedure takes a variable length argument list and
	   prints out the name of the function that failed, the
	   error message, usage if asked for, and then exits with
	   the code entered. */

{
#if !__STDC__
    int         status;                               /* status to exit with */
#endif
    va_list     args;           /* see vprintf(3) and varargs(5) for details */
    int         usage;                                 /* do you print usage */
    char      * fmt;                                        /* format string */

  fflush ( stdout );
#if __STDC__
  va_start ( args, status );
#else
  va_start ( args );
  status = va_arg ( args, int );         /* gets the first argument and type */
#endif
  usage = va_arg ( args, int );         /* gets the second argument and type */
  fmt = va_arg ( args, char * );
  fprintf ( stderr, ">> FATAL ERROR in %s:\n   ", progname );
  vfprintf ( stderr, fmt, args );                 /* print out error message */

  if ( usage )
    print_usage ();

  va_end ( args );
  exit ( status );
}                                                                   /* uquit */

